node/src/reactor/joiner.rs

```diff
@@ -485,8 +490,20 @@ impl reactor::Reactor for Reactor {
             .map(|(pk, motes)| (pk, motes.value()))
             .collect();
 
-        let linear_chain_sync =
-            LinearChainSync::new(registry, init_hash, validator_weights.clone())?;
+        let linear_chain_sync = {
+            let maybe_state =
+                linear_chain_sync::read_init_state(&storage, chainspec_loader.chainspec())?;
+            if let Some(state) = maybe_state {
+                LinearChainSync::from_state(registry, chainspec_loader.chainspec(), state)?
```

Comment:
Cleaner to just have `LinearChainSync::new` take the chainspec and storage and do this there.

Fast sync needs storage anyway, it's better if we keep the APIs of the two implementations as close as is reasonably possible.

Reply:
Done in [`087dddd` (#904)](https://github.com/CasperLabs/casper-node/pull/904/commits/087dddd78b6b23012eca2d7f765adf874623c248).

Reply:
> Fast sync needs storage anyway, it's better if we keep the APIs of the two implementations as close as is reasonably possible.


Actually, I didn't want to change the common API precisely b/c I don't know if fast sync needs the same arguments :+1: